src/main/java/com/urbanairship/datacube/dbharnesses/HBaseDbHarness.java

```diff
@@ -268,6 +296,46 @@ private long increment(byte[] rowKey, T op) throws IOException {
         return WithHTable.increment(pool, tableName, rowKey, cf, QUALIFIER, amount);
     }
 
+    /**
+     * @param writes map from rowkey to the operation (which had better be bytes compatible with long)
+     *
+     * @return A map from bytes to the new values written to the database.
+     *
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public Map<BoxedByteArray, byte[]> increment(Map<BoxedByteArray, T> writes) throws IOException, InterruptedException {
+        ImmutableMap.Builder<BoxedByteArray, byte[]> successes = ImmutableMap.builder();
+
+        List<Row> increments = new ArrayList<>();
+
+        List<Map.Entry<BoxedByteArray, byte[]>> entries = writes.entrySet().stream()
+                .map(e -> new AbstractMap.SimpleEntry<BoxedByteArray, byte[]>(e.getKey(), e.getValue().serialize()))
+                .collect(Collectors.toList());
+
+
+        for (Map.Entry<BoxedByteArray, byte[]> entry : entries) {
+            long amount = Bytes.toLong(entry.getValue());
+            Increment increment = new Increment(entry.getKey().bytes);
+            increment.addColumn(cf, QUALIFIER, amount);
+            incrementSize.update(amount);
+            increments.add(increment);
+        }
+
+        Object[] objects = WithHTable.batch(pool, tableName, increments);
+
+        for (int i = 0; i < objects.length; ++i) {
+            if (objects[i] != null) {
+                Result result = (Result) objects[i];
+                byte[] value = result.getValue(cf, QUALIFIER);
+                successes.put(entries.get(i).getKey(), value);
+            }
+        }
+
+        return successes.build();
```

Comment:
Should we retry failed writes here? We've definitely seen errors due to too many concurrent increments that'd cause partial failures.

Reply:
Added (weirdly) by throwing IO in https://github.com/urbanairship/datacube/pull/86/commits/cc437d425a1ff9f5931f7af3351585785202f9d2

Reply:
(although my personal preference would be to make the FlushWorkerRunnable a callable which returns the failed puts, i'm hesitant to do that in this change set)

Reply:
That's what I did in the "If I were rewriting datacube today how would I do it" branch I have... somewhere

Reply:
I have so many of those branches. :(